lib/SILOptimizer/Mandatory/MandatoryCombiner.cpp

```diff
@@ -0,0 +1,181 @@
+
+    auto fasArgs = fas.getArguments();
+    if (!areAllValuesTrivial(fasArgs, *function)) {
+      continue;
+    }
+
+    SmallVector<SILValue, 8> argsVec;
+    llvm::copy(paiArgs, std::back_inserter(argsVec));
+    llvm::copy(fasArgs, std::back_inserter(argsVec));
+
+    SILBuilderWithScope builder(fasi);
+    ApplyInst *appInst = builder.createApply(
+        /*Loc=*/fasi->getDebugLocation().getLocation(), /*Fn=*/callee,
+        /*Subs=*/pai->getSubstitutionMap(),
+        /*Args*/ argsVec,
+        /*isNonThrowing=*/fasi->isNonThrowing(),
+        /*SpecializationInfo=*/pai->getSpecializationInfo());
+    fasi->replaceAllUsesWith(appInst);
+    fasi->eraseFromParent();
+    erasedFullApply = true;
+  }
+  bool deletedDeadClosure = tryDeleteDeadClosure(pai);
+  return {erasedFullApply, deletedDeadClosure};
+}
+
+//===----------------------------------------------------------------------===//
+//                            Top Level Entrypoint
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+struct MandatoryCombiner : SILFunctionTransform {
+  void run() override {
+    bool madeChange = false;
+    auto *f = getFunction();
+    for (auto &block : *f) {
+      for (auto ii = block.begin(), ie = block.end(); ii != ie;) {
+        auto *inst = &*ii;
+        // If any action is taken, the current instruction will be deleted.
+        // That instruction would be the definition of a partial apply.  Because
+        // the definition dominates all its uses, the previous instruction will
+        // be unaffected by the removal of the instruction and its uses.  So
+        // move the iterator back a step.  If no action is taken, it will be
+        // advanced twice.  If an action is taken, it will be advanced once to a
+        // new instruction (since the current one will have been deleted).
+        ii = prev_or_default(ii, block.begin(), block.end());
+
+        bool deleted = false;
+        if (auto *pai = dyn_cast<PartialApplyInst>(inst)) {
+          auto result = processPartialApply(pai);
+          deleted = result.second;
+          madeChange |= result.first || result.second;
+        }
+
+        if (deleted) {
+          // The deletion succeeded.  The iterator has already been moved back
+          // a step so as to avoid invalidation.  Advancing it one step will
+          // not advance it to the current instruction since that has been
+          // deleted.  Instead, it will advance it to the next *remaining*
+          // instruction after the previous instruction as desired.
+          if (ii != block.end()) {
+            ii = next_or_end(ii, block.end());
+          } else {
+            ii = block.begin();
+          }
+          continue;
+        }
+
+        // No action was taken.  The iterator has already been moved back a
+        // step.  Consequently it is insufficient to advance it once--it must
+        // be advanced twice.
+        if (ii != block.end()) {
```

Comment:
I think that you can simplify this code (lines 163-167) as well.

Consider a situation where we were not processing the first instruction (so we know that ii != block.end()). In that case, we should just be able to do:

```
++ii;
++ii;
```

Now lets consider the case where we did have the first element (and so ii == block.end()). In that case, we just need to recognize the block.end() and handle that specially, but the 2nd increment does not need to do anything special. E.x.:

```
ii = next_or_default(ii, block.end() /*comparison*/, block.begin() /*default*/);
++ii;
```

Notice that the first next_or_default trivially also handles the first case since we know that ii != block.end(). I think that suggests that we can just use the 2nd e.x. here. So you can just use that and get rid of the conditional code.

@nate-chandler your thoughts?

Reply:
Makes sense.  We should be able to simplify a bit further: in the case where `ii == block.end()` we know what `next_or_default(ii, block.end(), block.begin())` will be, namely `block.begin()`:

``
ii = block.begin();
++ii;
``

Thoughts?

Reply:
Nevermind, I missed the last paragraph where you suggested merging the two branches.

Reply:
Fixed.